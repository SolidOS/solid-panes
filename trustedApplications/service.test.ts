/* eslint-env jest */
const $rdf = require('rdflib')
const ns = require('solid-namespace')($rdf)
const { getStatementsToDelete, getStatementsToAdd, deserialiseMode } = require('./service')

describe('getStatementsToDelete', () => {
  it('should return an empty array when there are no statements', () => {
    const mockStore = $rdf.graph()
    const mockOrigin = $rdf.sym('https://fake.origin')
    const mockProfile = $rdf.sym('https://fake.profile#me')
    expect(getStatementsToDelete(mockOrigin, mockProfile, mockStore, ns)).toEqual([])
  })

  it('should return all statements for the given origin', () => {
    const mockStore = $rdf.graph()
    const mockApplication = $rdf.sym('https://fake.app')
    const mockOrigin = $rdf.sym('https://fake.origin')
    const mockProfile = $rdf.sym('https://fake.profile#me')
    mockStore.add(mockApplication, ns.acl('origin'), mockOrigin)
    mockStore.add(mockApplication, ns.acl('mode'), ns.acl('Read'))
    mockStore.add(mockProfile, ns.acl('trustedApp'), mockApplication)
    const statementsToDelete = getStatementsToDelete(mockOrigin, mockProfile, mockStore, ns)
    expect(statementsToDelete.length).toBe(3)
    expect(statementsToDelete).toMatchSnapshot()
  })

  it('should not return statements for a different origin', () => {
    const mockStore = $rdf.graph()
    const mockApplication = $rdf.sym('https://fake.app')
    const mockOrigin = $rdf.sym('https://fake.origin')
    const mockProfile = $rdf.sym('https://fake.profile#me')
    mockStore.add(mockApplication, ns.acl('origin'), mockOrigin)
    mockStore.add(mockApplication, ns.acl('mode'), ns.acl('Read'))
    mockStore.add(mockProfile, ns.acl('trustedApp'), mockApplication)

    const statementsToDelete = getStatementsToDelete($rdf.lit('A different origin'), mockProfile, mockStore, ns)
    expect(statementsToDelete.length).toBe(0)
    expect(statementsToDelete).toEqual([])
  })
})

describe('getStatementsToAdd', () => {
  it('should return all required statements to add the given permissions for a given origin', () => {
    const mockOrigin = $rdf.sym('https://fake.origin')
    const mockProfile = $rdf.sym('https://fake.profile#me')
    const modes = ['read', 'write']

    const statementsToAdd = getStatementsToAdd(mockOrigin, 'mock_app_id', modes, mockProfile, ns)
    expect(statementsToAdd.length).toBe(4)
    expect(statementsToAdd).toMatchSnapshot()
  })
})

describe('deserialiseMode', () => {
  it('should convert a full namespaced ACL to a plaintext string', () => {
    expect(deserialiseMode($rdf.sym(ns.acl('read')), ns)).toBe('read')
    expect(deserialiseMode($rdf.sym(ns.acl('append')), ns)).toBe('append')
    expect(deserialiseMode($rdf.sym(ns.acl('write')), ns)).toBe('write')
    expect(deserialiseMode($rdf.sym(ns.acl('control')), ns)).toBe('control')
  })
})
