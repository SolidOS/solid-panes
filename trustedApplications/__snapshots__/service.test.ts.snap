// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStatementsToAdd should return all required statements to add the given permissions for a given origin 1`] = `
Array [
  Statement {
    "object": BlankNode {
      "id": "bn_mock_app_id",
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#trustedApp",
    },
    "subject": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile#me",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile",
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.origin",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#origin",
    },
    "subject": BlankNode {
      "id": "bn_mock_app_id",
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile",
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#read",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#mode",
    },
    "subject": BlankNode {
      "id": "bn_mock_app_id",
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile",
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#write",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#mode",
    },
    "subject": BlankNode {
      "id": "bn_mock_app_id",
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile",
    },
  },
]
`;

exports[`getStatementsToDelete should return all statements for the given origin 1`] = `
Array [
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.app",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#trustedApp",
    },
    "subject": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile#me",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "chrome:theSession",
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.origin",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#origin",
    },
    "subject": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.app",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "chrome:theSession",
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#Read",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#mode",
    },
    "subject": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.app",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "chrome:theSession",
    },
  },
]
`;
